system/smbus.c

```diff
@@ -1189,35 +1119,220 @@ static spd_info parse_spd_sdram(uint8_t smb_idx, uint8_t slot_idx)
+    0x1D71, // Patsburg (PCH) IDF
+    0x1D72, // Patsburg (PCH) IDF
+    0x2330, // DH89xxCC (PCH)
+    0x1E22, // Panther Point (PCH)
+    0x8C22, // Lynx Point (PCH)
+    0x9C22, // Lynx Point-LP (PCH)
+    0x1F3C, // Avoton (SOC)
+    0x8D22, // Wellsburg (PCH)
+    0x8D7D, // Wellsburg (PCH) MS
+    0x8D7E, // Wellsburg (PCH) MS
+    0x8D7F, // Wellsburg (PCH) MS
+    0x23B0, // Coleto Creek (PCH)
+    0x8CA2, // Wildcat Point (PCH)
+    0x9CA2, // Wildcat Point-LP (PCH)
+    0x0F12, // BayTrail (SOC)
+    0x2292, // Braswell (SOC)
+    0xA123, // Sunrise Point-H (PCH)
+    0x9D23, // Sunrise Point-LP (PCH)
+    0x19DF, // Denverton  (SOC)
+    0x1BC9, // Emmitsburg (PCH)
+    0xA1A3, // Lewisburg (PCH)
+    0xA223, // Lewisburg Super (PCH)
+    0xA2A3, // Kaby Lake (PCH-H)
+    0x31D4, // Gemini Lake (SOC)
+    0xA323, // Cannon Lake-H (PCH)
+    0x9DA3, // Cannon Lake-LP (PCH)
+    0x18DF, // Cedar Fork (PCH)
+    0x34A3, // Ice Lake-LP (PCH)
+    0x38A3, // Ice Lake-N (PCH)
+    0x02A3, // Comet Lake (PCH)
+    0x06A3, // Comet Lake-H (PCH)
+    0x4B23, // Elkhart Lake (PCH)
+    0xA0A3, // Tiger Lake-LP (PCH)
+    0x43A3, // Tiger Lake-H (PCH)
+    0x4DA3, // Jasper Lake (SOC)
+    0xA3A3, // Comet Lake-V (PCH)
+    0x7AA3, // Alder Lake-S (PCH)
+    0x51A3, // Alder Lake-P (PCH)
+    0x54A3, // Alder Lake-M (PCH)
+};
+
+static bool find_in_did_array(uint16_t did, const uint16_t * ids, unsigned int size) {
+    for (unsigned int i = 0; i < size; i++) {
+        if (*ids++ == did) {
+            return true;
+        }
+    }
+    return false;
+}
+
+
+static bool find_smb_controller(uint16_t vid, uint16_t did) {
+
+    switch(vid)
+    {
+        case VID_INTEL:
+        {
+            if (find_in_did_array(did, intel_ich5_dids, sizeof(intel_ich5_dids) / sizeof(intel_ich5_dids[0]))) {
+                return ich5_get_smb_base();
+            }
+/*
+            if (   did == 0x7113 // 82371AB/EB/MB PIIX4
+                || did == 0x719B // 82440/82443MX PMC
+               ) {
+                return piix4_get_smb_base();
+            }
+*/
+          // 0x0C59-0x0C5E, 0x18AC, 0x19AC, 0x1BFF, 0x1F15, 0x1F3D Atom S1200 / C2000 / C3000 SMBus controllers may not be SPD-related.
+            return false;
+        }
+
+        case VID_AMD:
+            switch(did)
+            {
+/*
+                case 0x740B: // AMD756
+                case 0x7413: // AMD766
+                case 0x7443: // AMD768
+                case 0x746B: // AMD8111_SMBUS
+                    return amd756nforce_get_smb_base();
+                case 0x746A: // AMD8111_SMBUS2
+                    return amd8111_get_smb_base();
+                case 0x780B: // FCH SMBus Controller
+                    return piix4_get_smb_base();
+*/
+                case 0x790B: // AMD FCH (Zen 2/3)
+                    return fch_zen_get_smb_base();
+                default:
+                return false;
+            }
+            break;
 
-    return -1;
+        case VID_HYGON:
```

Comment:
Hygon's CPU are strictly identical to AMD's CPU (at least from a SMBus PoV) as they're only rebranded EPYC with some features fused-down, so this can be handled on the same case.